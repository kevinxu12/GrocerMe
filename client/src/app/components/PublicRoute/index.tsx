/**
 * @file Route for already loggedin users. If a user is logged in, should redirect to user home page
 * @author Kevin Xu
 */
import React from 'react';
import { useSelector } from 'react-redux';
import { Route, Redirect } from 'react-router-dom';
import { RootState } from 'types';
import { Constants } from 'utils/constants';

/**
 * @param {...any} root0 props to pass to component if authenticated
 * @param {React.ElementType} root0.component component to return if authenticated
 * @returns {React.ElementType} either a Redirect or the component passed in props
 */
const PublicRoute = ({ component: Component, ...rest }) => {
  const isAuthenticated = useSelector((state: RootState) =>
    state ? state.auth.username : true,
  );
  return (
    <Route
      {...rest}
      render={props =>
        isAuthenticated ? (
          <Redirect to={Constants.USER_HOME} />
        ) : (
          <Component {...props} />
        )
      }
    />
  );
};

export default PublicRoute;
