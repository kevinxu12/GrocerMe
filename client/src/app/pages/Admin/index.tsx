/**
 * @file Admin Dashboard
 * @author Kevin Xu
 */
import React from 'react';

/**
 * Render a dummy Admin Dashboard Component
 *
 * @returns {React.ReactElement} Admin Dashboard Component
 */
export const AdminDashboard = (): React.ReactElement => {
  return <div> Placeholder. Please make a dashboard eventually </div>;
};
