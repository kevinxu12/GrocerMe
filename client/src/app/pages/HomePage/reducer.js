/*
 * HomeReducer
 *
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 *
 */

import produce from 'immer';
import { Constants } from 'utils/constants';
import { CHANGE_USERNAME } from './constants';

// The initial state of the App
// const storagedata = JSON.parse(localStorage.getItem("data"));
export const initialState = {
  username: Constants.DEFAULT_STRING,
};

/* eslint-disable default-case, no-param-reassign */
const homeReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case CHANGE_USERNAME:
        // Delete prefixed '@' from the github username
        draft.username = action.username;
        break;
    }
  });

export default homeReducer;
