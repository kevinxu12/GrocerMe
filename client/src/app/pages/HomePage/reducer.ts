/*
 * HomeReducer
 *
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 *
 */

import produce from 'immer';
import { HomeAction } from 'types/actions';
import { HomeState } from 'types/RootState';
import { Constants } from 'utils/constants';
import { CHANGE_USERNAME } from './constants';

// The initial state of the App
export const initialState: HomeState = {
  username: Constants.DEFAULT_STRING,
};

/* eslint-disable default-case, no-param-reassign */
const homeReducer = (
  state: HomeState = initialState,
  action: HomeAction,
): HomeState => {
  return produce(state, draft => {
    switch (action.type) {
      case CHANGE_USERNAME:
        draft.username = action.payload.username;
        break;
    }
  });
};

export default homeReducer;
