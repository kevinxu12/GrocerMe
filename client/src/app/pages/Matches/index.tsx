/**
 * @file Frontend for Matches
 * @author Kevin Xu
 */

import {
  Card,
  CardActions,
  CardContent,
  Link,
  Stack,
  Typography,
} from '@mui/material';
import React, { useEffect, useState } from 'react';
import { ItemMatchRequest, RequestStatus } from 'types/rest';
import Api from 'utils/api';
import { Constants } from 'utils/constants';
import { parseAxiosSuccessResponse } from 'utils/request';
/**
 *
 */
export const Home = (): React.ReactElement => {
  const [matches, setMatches] = useState<ItemMatchRequest[]>([]);
  useEffect(() => {
    /**
     *
     */
    const fetchMatchesAsSupplier = async () => {
      const response = await Api.get('/itemMatchRequestsForUser', {
        status: RequestStatus.ACCEPTED,
      });
      const itemMatchRequests =
        parseAxiosSuccessResponse<ItemMatchRequest[]>(response);
      if (itemMatchRequests) {
        setMatches(itemMatchRequests);
      }
    };
    fetchMatchesAsSupplier();
  }, []);
  if (matches.length === 0) {
    return <div>No Matches for Now: Stay put</div>;
  }
  return (
    <div>
      <Stack spacing={2}>
        <Typography> All Matches as a Supplier </Typography>
        {matches.map((match: ItemMatchRequest, index: number) => (
          <Card key={index} sx={{ display: 'flex', width: '50%' }}>
            <CardContent>
              <Typography variant="body2" sx={{ mb: 1.5 }} component="div">
                Match between {match.requesterEmail} and {match.supplierEmail}
              </Typography>
            </CardContent>
            <CardActions>
              <Link href={Constants.MATCH + '/' + match._id}>View Match</Link>
            </CardActions>
          </Card>
        ))}
      </Stack>
    </div>
  );
};
