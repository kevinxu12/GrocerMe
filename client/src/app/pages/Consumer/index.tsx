/**
 * @file A Home page for logged-in users, both consumers and suppliers
 * Dashboard is not a private route, because dashboard will refresh periodically to check if the user is still logged in or not.
 * @author Kevin Xu
 */
import React, { ComponentType, Dispatch, memo } from 'react';
import { AxiosResponse } from 'axios';
import changeAuth, { logout } from 'store/auth/actions';
import { Role, User } from 'types/rest';
import { connect } from 'react-redux';
import { compose } from '@reduxjs/toolkit';
import { useHistory } from 'react-router-dom';
import { ChangeAuthAction, LogoutAction } from 'types/actions';
import { createStructuredSelector } from 'reselect';
import { makeSelectRoles, makeSelectUsername } from 'store/auth/selectors';
import useDeepCompareEffect from 'use-deep-compare-effect';
import { Wrapper } from 'app/components/PrivateWrapper';
import { useSocket } from 'context/SocketContext';
import { Api, parseAxiosSuccessResponse } from 'utils/request';

interface DashboardPropsType {
  changeAuth: (email: string, roles?: Role[]) => void;
  logout: () => void;
  username: string;
  roles: Role[];
}

/**
 * Generate a dashboard React element. This should be the landing page for a logged-in, non-first-time user
 *
 * @param {object} props props passed
 * @param {Function} props.changeAuth check if logged in, if so store email into redux
 * @param {Function} props.logout logout the user from redux
 * @param {string} props.username the username of logged-in user in redux, if this exists
 * @param {Role[]} props.roles the roles of logged-in user in redux, if this exists
 * @returns {React.ReactElement} Dashboard Component
 */
const Dashboard = ({
  changeAuth,
  logout,
  username,
  roles,
}: DashboardPropsType): React.ReactElement => {
  const history = useHistory();
  const socket = useSocket();
  useDeepCompareEffect(() => {
    /**
     * Test function to listen to a socket thats emitted
     */
    async function checkLogin(): Promise<void> {
      const response: AxiosResponse = await Api.get('/loggedin');
      const user = parseAxiosSuccessResponse<User>(response);
      console.log(user);
      if (user == null) {
        logout();
        history.push('/');
        return;
      }
      changeAuth(user.email, user.roles);
      socket.emit('login', user);
    }
    checkLogin();
  }, [username, roles]);
  return <Wrapper>Sample Dashboard</Wrapper>;
};

/**
 * Maps redux store values to component props
 */
const mapStateToProps = createStructuredSelector({
  username: makeSelectUsername(),
  roles: makeSelectRoles(),
});

/**
 * Maps dispatch functions to component props
 *
 * @param {Dispatch} dispatch the dispatch object from redux
 * @returns {object} Object passed to props containing redux dispatch functions
 */
function mapDispatchToProps(
  dispatch: Dispatch<ChangeAuthAction | LogoutAction>,
) {
  return {
    /**
     * Stores info of logged-in user in redux
     *
     * @param {string} email the email of the logged-in user
     * @param {Role[]} roles the optional-roles of the logged-in user
     */
    changeAuth: (email: string, roles?: Role[]) => {
      dispatch(changeAuth({ username: email, roles }));
    },
    /**
     * Logs out a user if no email returned from redux (session object is expired)
     */
    logout: () => {
      dispatch(logout());
    },
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose<ComponentType>(withConnect, memo)(Dashboard);
