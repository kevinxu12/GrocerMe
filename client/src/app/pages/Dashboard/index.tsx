/**
 * @file A Home page for logged-in users, both consumers and suppliers
 * @author Kevin Xu
 */
import React, { ComponentType, Dispatch, memo } from 'react';
import axios, { AxiosResponse } from 'axios';
import changeAuth, { logout } from 'store/auth/actions';
import { Role, StatusCode, User } from 'types/rest';
import { connect } from 'react-redux';
import { compose } from '@reduxjs/toolkit';
import { useHistory } from 'react-router-dom';
import generateServerUrl from 'utils/url';
import { ChangeAuthAction, LogoutAction } from 'types/actions';
import { createStructuredSelector } from 'reselect';
import { makeSelectRoles, makeSelectUsername } from 'store/auth/selectors';
import useDeepCompareEffect from 'use-deep-compare-effect';
import { Wrapper } from 'app/components/PrivateWrapper';

interface DashboardPropsType {
  changeAuth: (email: string, roles?: Role[]) => void;
  logout: () => void;
  username: string;
  roles: Role[];
}

/**
 * Generate a dashboard React element. This should be the landing page for a logged-in, non-first-time user
 *
 * @param {object} props props passed
 * @param {Function} props.changeAuth check if logged in, if so store email into redux
 * @param {Function} props.logout logout the user from redux
 * @param {string} props.username the username of logged-in user in redux, if this exists
 * @param {Role[]} props.roles the roles of logged-in user in redux, if this exists
 * @returns {React.ReactElement} Dashboard Component
 */
const Dashboard = (props: DashboardPropsType): React.ReactElement => {
  const history = useHistory();
  useDeepCompareEffect(() => {
    /**
     * Test function to listen to a socket thats emitted
     */
    async function checkLogin(): Promise<void> {
      const response: AxiosResponse = await axios.get(
        generateServerUrl('/loggedIn'),
        { withCredentials: true },
      );
      const responseBody: AxiosResponse['data'] = response.data;
      if (responseBody.statusCode === StatusCode.SUCCESS) {
        const user: User = responseBody.data;
        if (user) {
          console.log('changing props');
          props.changeAuth(user.email, user.roles);
        } else {
          props.logout();
          history.push('/');
        }
      }
    }
    checkLogin();
  }, [props.username, props.roles]);
  return <Wrapper>Sample Dashboard</Wrapper>;
};

/**
 * Maps redux store values to component props
 */
const mapStateToProps = createStructuredSelector({
  username: makeSelectUsername(),
  roles: makeSelectRoles(),
});

/**
 * Maps dispatch functions to component props
 *
 * @param {Dispatch} dispatch the dispatch object from redux
 * @returns {object} Object passed to props containing redux dispatch functions
 */
function mapDispatchToProps(
  dispatch: Dispatch<ChangeAuthAction | LogoutAction>,
) {
  return {
    /**
     * Stores info of logged-in user in redux
     *
     * @param {string} email the email of the logged-in user
     * @param {Role[]} roles the optional-roles of the logged-in user
     */
    changeAuth: (email: string, roles?: Role[]) => {
      dispatch(changeAuth({ username: email, roles }));
    },
    /**
     * Logs out a user if no email returned from redux (session object is expired)
     */
    logout: () => {
      dispatch(logout());
    },
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose<ComponentType>(withConnect, memo)(Dashboard);
