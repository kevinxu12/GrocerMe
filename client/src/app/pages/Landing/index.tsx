/**
 * @file Component for HomePage
 * We shoud move the landing / home page out using Material UI instead of Hardcoded Components
 * @author Kevin Xu
 */
import React from 'react';
import styled from 'styled-components/macro';

import { PageWrapper } from 'app/components/PageWrapper';
import { StyledConstants } from 'styles/StyleConstants';
import OAuthButton from 'app/components/OAuthButton';
import { Constants } from 'utils/constants';

/**
 * @returns {React.FC} Homepage Component
 */
export function HomePage() {
  return (
    <PageWrapper>
      <OverlapGroup3>
        <OverlapGroup1>
          <BlobShape1>
            <img loading="lazy" src={Constants.BLOB_PNG_URL} alt="IMG" />
          </BlobShape1>
        </OverlapGroup1>
        <Heading>
          <Title>Need Groceries?</Title>
          <Description>
            Buy Groceries Quick, and get them even quicker!
          </Description>
        </Heading>
        <OverlapGroup2>
          <OAuthButton />
        </OverlapGroup2>
      </OverlapGroup3>
    </PageWrapper>
  );
}

const OverlapGroup3 = styled.div`
  width: 100%;
  height: 80%;
  position: relative;
  margin-left: 3%;
`;

const OverlapGroup1 = styled.div`
  width: 100%;
  top: 0;
  left: 0;
  display: flex;
  flex-direction: column;
  padding: 3% 0;
  align-items: flex-start;
  min-height: 80%;
`;

const BlobShape1 = styled.div`
  width: 48%;
  align-self: flex-end;
  margin-top: 5%;
  margin-right: 3%;
  display: flex;
  align-items: flex-start;
  overflow: hidden;
  selection-color: blue;
`;

const Heading = styled.div`
  position: absolute;
  width: 50%;
  height: 20%;
  top: 30%;
  left: 0;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
`;

const Title = styled.h1`
  min-height: 50%;
  font-weight: 400;
  color: ${p => p.theme.grey};
  font-size: ${StyledConstants.FONT_SIZE_XXL};
  letter-spacing: 0;
  line-height: 100%;
  white-space: nowrap;
`;

const Description = styled.div`
  min-height: 40%;
  margin-top: 5;
  font-weight: 400;
  color: #66645e;
  font-size: ${StyledConstants.FONT_SIZE_L};
  letter-spacing: 0;
  line-height: 5%;
`;

const OverlapGroup2 = styled.div`
  position: absolute;
  width: 100%;
  height: 100%;
  top: 60%;
  left: 0;
`;
