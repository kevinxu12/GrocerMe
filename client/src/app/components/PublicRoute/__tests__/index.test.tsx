
import * as React from 'react';
import { initialState } from 'store/auth/reducer';
import { HomePage } from 'app/pages/Landing';
import PublicRoute from '..';
import { generateAuthState, MockStoreWrapper } from 'utils/test';
import { RootState } from 'types';
import * as authHelper from 'utils/auth';
jest.mock('utils/auth');

describe('<PublicRoute  />', () => {
    const mockStoreWrapper = new MockStoreWrapper<RootState>({auth: initialState});
  
    it('No logged in user should stay at the home page', () => {
      const auth = initialState;
      const pr = mockStoreWrapper.renderComponentWithRouterProvider(<PublicRoute auth = {auth} component = {HomePage} />)
      expect(pr.container.firstChild).toMatchSnapshot();
      expect(authHelper.isRoleCodeIncluded).toBeCalledTimes(1);
    });
    it('Logged in user with admin roles should move to admin', () => {

      const auth = generateAuthState([]);
      const pr = mockStoreWrapper.renderComponentWithRouterProvider(<PublicRoute auth = {auth} component = {HomePage} />, {auth})
      expect(pr.container.firstChild).toMatchSnapshot();
      expect(authHelper.isRoleCodeIncluded).toBeCalledTimes(1);
    });
    it('Logged in user with consuer roles should move to consumer home', () => {
      const auth = generateAuthState([]);
      const pr = mockStoreWrapper.renderComponentWithRouterProvider(<PublicRoute auth = {auth} component = {HomePage} />, {auth})
      expect(pr.container.firstChild).toMatchSnapshot();
      expect(authHelper.isRoleCodeIncluded).toBeCalledTimes(1);
    });
  });