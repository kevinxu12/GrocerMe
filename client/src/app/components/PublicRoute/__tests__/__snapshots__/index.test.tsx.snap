// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PublicRoute  /> Logged in user with admin roles should move to admin 1`] = `null`;

exports[`<PublicRoute  /> Logged in user with consuer roles should move to consumer home 1`] = `null`;

exports[`<PublicRoute  /> No logged in user should stay at the home page 1`] = `
.c0 {
  width: 1400px;
  margin: 0 auto;
  padding: 0 1.5rem;
  box-sizing: content-box;
}

.c1 {
  width: 100%;
  height: 80%;
  position: relative;
  margin-left: 3%;
}

.c2 {
  width: 100%;
  top: 0;
  left: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 3% 0;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  min-height: 80%;
}

.c3 {
  width: 48%;
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
  margin-top: 5%;
  margin-right: 3%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  overflow: hidden;
  selection-color: blue;
}

.c4 {
  position: absolute;
  width: 50%;
  height: 20%;
  top: 30%;
  left: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c5 {
  min-height: 50%;
  font-weight: 400;
  font-size: 64px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 100%;
  white-space: nowrap;
}

.c6 {
  min-height: 40%;
  margin-top: 5;
  font-weight: 400;
  color: #66645e;
  font-size: 24px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 5%;
}

.c7 {
  position: absolute;
  width: 100%;
  height: 100%;
  top: 60%;
  left: 0;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <img
          alt="IMG"
          src="https://i.ibb.co/Hp5G1G4/blob-shape-1.png"
        />
      </div>
    </div>
    <div
      class="c4"
    >
      <h1
        class="c5"
      >
        Need Groceries?
      </h1>
      <div
        class="c6"
      >
        Buy Groceries Quick, and get them even quicker!
      </div>
    </div>
    <div
      class="c7"
    >
      <div
        role="button"
        style="background-color: rgb(66, 133, 244); color: rgb(255, 255, 255); height: 50px; width: 240px; text-align: center; vertical-align: center; box-shadow: 0 2px 4px 0 rgba(0,0,0,.25); font-size: 16px; line-height: 48px; display: block; border-radius: 1px; transition: background-color .218s, border-color .218s, box-shadow .218s; font-family: Roboto,arial,sans-serif; cursor: pointer; user-select: none;"
        tabindex="0"
        type="dark"
      >
        <div
          style="width: 48px; height: 48px; text-align: center; vertical-align: center; display: block; margin-top: 1px; margin-left: 1px; float: left; background-color: rgb(255, 255, 255); border-radius: 1px; white-space: nowrap;"
        >
          <svg
            height="46px"
            style="width: 48px; height: 48px; display: block;"
            version="1.1"
            viewBox="0 0 46 46"
            width="46px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <defs>
              <filter
                filterUnits="objectBoundingBox"
                height="200%"
                id="filter-1"
                width="200%"
                x="-50%"
                y="-50%"
              >
                <feoffset
                  dx="0"
                  dy="1"
                  in="SourceAlpha"
                  result="shadowOffsetOuter1"
                />
                <fegaussianblur
                  in="shadowOffsetOuter1"
                  result="shadowBlurOuter1"
                  stdDeviation="0.5"
                />
                <fecolormatrix
                  in="shadowBlurOuter1"
                  result="shadowMatrixOuter1"
                  type="matrix"
                  values="0 0 0 0 0   0 0 0 0 0   0 0 0 0 0  0 0 0 0.168 0"
                />
                <feoffset
                  dx="0"
                  dy="0"
                  in="SourceAlpha"
                  result="shadowOffsetOuter2"
                />
                <fegaussianblur
                  in="shadowOffsetOuter2"
                  result="shadowBlurOuter2"
                  stdDeviation="0.5"
                />
                <fecolormatrix
                  in="shadowBlurOuter2"
                  result="shadowMatrixOuter2"
                  type="matrix"
                  values="0 0 0 0 0   0 0 0 0 0   0 0 0 0 0  0 0 0 0.084 0"
                />
                <femerge>
                  <femergenode
                    in="shadowMatrixOuter1"
                  />
                  <femergenode
                    in="shadowMatrixOuter2"
                  />
                  <femergenode
                    in="SourceGraphic"
                  />
                </femerge>
              </filter>
              <rect
                height="40"
                id="path-2"
                rx="2"
                width="40"
                x="0"
                y="0"
              />
              <rect
                height="38"
                id="path-3"
                rx="1"
                width="38"
                x="5"
                y="5"
              />
            </defs>
            <g
              fill="none"
              fill-rule="evenodd"
              id="Google-Button"
              stroke="none"
              stroke-width="1"
            >
              <g
                id="9-PATCH"
                transform="translate(-608.000000, -219.000000)"
              />
              <g
                id="btn_google_dark_normal"
                transform="translate(-1.000000, -1.000000)"
              >
                <g
                  filter="url(#filter-1)"
                  id="button"
                  transform="translate(4.000000, 4.000000)"
                >
                  <g
                    id="button-bg"
                  >
                    <use
                      fill="#4285F4"
                      fill-rule="evenodd"
                    />
                    <use
                      fill="none"
                    />
                    <use
                      fill="none"
                    />
                    <use
                      fill="none"
                    />
                  </g>
                </g>
                <g
                  id="button-bg-copy"
                >
                  <use
                    fill="#FFFFFF"
                    fill-rule="evenodd"
                  />
                  <use
                    fill="none"
                  />
                  <use
                    fill="none"
                  />
                  <use
                    fill="none"
                  />
                </g>
                <g
                  id="logo_googleg_48dp"
                  transform="translate(15.000000, 15.000000)"
                >
                  <path
                    d="M17.64,9.20454545 C17.64,8.56636364 17.5827273,7.95272727 17.4763636,7.36363636 L9,7.36363636 L9,10.845 L13.8436364,10.845 C13.635,11.97 13.0009091,12.9231818 12.0477273,13.5613636 L12.0477273,15.8195455 L14.9563636,15.8195455 C16.6581818,14.2527273 17.64,11.9454545 17.64,9.20454545 L17.64,9.20454545 Z"
                    fill="#4285F4"
                    id="Shape"
                  />
                  <path
                    d="M9,18 C11.43,18 13.4672727,17.1940909 14.9563636,15.8195455 L12.0477273,13.5613636 C11.2418182,14.1013636 10.2109091,14.4204545 9,14.4204545 C6.65590909,14.4204545 4.67181818,12.8372727 3.96409091,10.71 L0.957272727,10.71 L0.957272727,13.0418182 C2.43818182,15.9831818 5.48181818,18 9,18 L9,18 Z"
                    fill="#34A853"
                    id="Shape"
                  />
                  <path
                    d="M3.96409091,10.71 C3.78409091,10.17 3.68181818,9.59318182 3.68181818,9 C3.68181818,8.40681818 3.78409091,7.83 3.96409091,7.29 L3.96409091,4.95818182 L0.957272727,4.95818182 C0.347727273,6.17318182 0,7.54772727 0,9 C0,10.4522727 0.347727273,11.8268182 0.957272727,13.0418182 L3.96409091,10.71 L3.96409091,10.71 Z"
                    fill="#FBBC05"
                    id="Shape"
                  />
                  <path
                    d="M9,3.57954545 C10.3213636,3.57954545 11.5077273,4.03363636 12.4404545,4.92545455 L15.0218182,2.34409091 C13.4631818,0.891818182 11.4259091,0 9,0 C5.48181818,0 2.43818182,2.01681818 0.957272727,4.95818182 L3.96409091,7.29 C4.67181818,5.16272727 6.65590909,3.57954545 9,3.57954545 L9,3.57954545 Z"
                    fill="#EA4335"
                    id="Shape"
                  />
                  <path
                    d="M0,0 L18,0 L18,18 L0,18 L0,0 Z"
                    id="Shape"
                  />
                </g>
                <g
                  id="handles_square"
                />
              </g>
            </g>
          </svg>
        </div>
        <span>
          Sign in with Google
        </span>
      </div>
    </div>
  </div>
</div>
`;
